invoke/context.py

```diff
@@ -192,6 +202,134 @@ def sudo(self, command, **kwargs):
             else:
                 raise
 
+    def _prefix_commands(self, command):
+        """
+        Prefixes ``command`` with all prefixes found in ``command_prefixes``.
+
+        ``command_prefixes`` is a list of strings which is modified by the
+        `prefix` context manager.
+        """
+        prefixes = list(self.command_prefixes)
+        current_directory = self.cwd
+        if current_directory:
+            prefixes.insert(0, 'cd %s' % current_directory)
+
+        return ' && '.join(prefixes + [command])
+
+    @contextmanager
+    def prefix(self, command):
+        """
+        Prefix all nested `run`/`sudo` commands with given command plus ``&&``.
+
+        Most of the time, you'll want to be using this alongside a shell script
+        which alters shell state, such as ones which export or alter shell
+        environment variables.
+
+        For example, one of the most common uses of this tool is with the
+        ``workon`` command from `virtualenvwrapper
+        <https://virtualenvwrapper.readthedocs.io/en/latest/>`_::
+
+            with ctx.prefix('workon myvenv'):
+                ctx.run('./manage.py migrate')
+
+        In the above snippet, the actual shell command run would be this::
+
+            $ workon myvenv && ./manage.py migrate
+
+        This context manager is compatible with `cd`, so if your virtualenv
+        doesn't ``cd`` in its ``postactivate`` script, you could do the
+        following::
+
+            with ctx.cd('/path/to/app'):
+                with ctx.prefix('workon myvenv'):
+                    ctx.run('./manage.py migrate')
+                    ctx.run('./manage.py loaddata fixture')
+
+        Which would result in executions like so::
+
+            $ cd /path/to/app && workon myvenv && ./manage.py migrate
+            $ cd /path/to/app && workon myvenv && ./manage.py loaddata fixture
+
+        Finally, as alluded to above, `prefix` may be nested if desired, e.g.::
+
+            with ctx.prefix('workon myenv'):
+                ctx.run('ls')
+                with ctx.prefix('source /some/script'):
+                    ctx.run('touch a_file')
+
+        The result::
+
+            $ workon myenv && ls
+            $ workon myenv && source /some/script && touch a_file
+
+        Contrived, but hopefully illustrative.
+        """
+        self.command_prefixes.append(command)
+        yield
+        self.command_prefixes.pop()
+
+    @property
+    def cwd(self):
+        """
+        Returns the current working directory computed from invocations of the
+        `cd` context manager.
+        """
+        if not self.command_cwds:
+            return ''
+
+        # get the index for the subset of paths starting with the last / or ~
+        for i, path in reversed(list(enumerate(self.command_cwds))):
+            if path.startswith('~') or path.startswith('/'):
+                break
+
+        paths = [path.replace(' ', '\ ') for path in self.command_cwds[i:]]
```

Comment:
Forget offhand but we may want to see if there's a stronger "escape this path" function somewhere we can reuse - think escaping tildes or slashes in filenames and dumb shit like that.